File: flowersec-ts/src/v4/rekeyReference.test.ts
// Public fixed transactions only. Expected-byte matching is not a receiver
// codec, barrier proof, epoch installation, clock or single-publisher gate.
import { readFileSync } from "node:fs";
import { expect, it } from "vitest";
import { x25519 } from "@noble/curves/ed25519.js";
import { p256 } from "@noble/curves/nist.js";
import { expand } from "@noble/hashes/hkdf.js";
import { hmac } from "@noble/hashes/hmac.js";
import { sha256 } from "@noble/hashes/sha2.js";
import { chacha20poly1305 } from "@noble/ciphers/chacha.js";
import { gcm } from "@noble/ciphers/aes.js";
import { transportV4Domains, transportV4RekeyRegistry, transportV4RecordRegistry, transportV4Registry } from "../generated/transportV4Registry.js";

type Bytes = Uint8Array;
type Values = Record<string, Bytes | bigint | Values[]>;
type Context = { profile: string; handshake_hash_hex: string; context_digest_hex: string; epoch: number; next_epoch: number; rekey_id_hex: string };
type Phase = { id: string; schema: string; phase: number; role: number; base_hex: string; message_hex: string;
  unsigned_hex: string; key_info_hex: string; key_hex: string; mac_message_hex: string; confirmation_mac_hex: string;
  expected: Record<string, string>; record: { epoch: number; direction: number; sequence: string; wire_hex: string } };
type Round = { id: string; context: Context; input: {client_private_hex: string; server_private_hex: string;
  client_barrier: Record<string, string>[]; server_barrier: Record<string, string>[]; client_old_sequence: string; server_old_sequence: string};
  old_root_hex: string; new_root_hex: string; secret_hex: string; dh_hex: string; prk_hex: string;
  init_digest_hex: string; transcript_hex: string; phases: Phase[] };
type Field = { name: string; type: string; length?: number; const?: number; profile_public_key?: boolean; items?: {schema_ref: string} };
type Definition = { sender_role: number; mac_field: number; fields: Record<string, Field> };
type Layout = readonly {name: string; type: string; const?: number; max?: number}[];
type Part = { name: string; encoding: string; length?: number };
const reg = transportV4RekeyRegistry as unknown as Record<string, Definition>;
const records = transportV4RecordRegistry as { envelope: {layout: Layout}; header: Layout; nonce: Layout; profiles: Record<string, {dh_algorithm: number; dh_public_bytes: number; record_aead: string; tag_bytes: number}> };
const domains = transportV4Domains as unknown as {name: string; label_bytes: string; input_schema: {parts: Part[]}}[];
const corpus = JSON.parse(readFileSync(new URL("../../../testdata/transport_v4/rekey.json", import.meta.url), "utf8")) as {
  schema_sha256: string; rounds: Round[]; negatives: {id: string; source: string; message_hex: string; base_hex: string; context_patch: Partial<Context>; expected: Record<string,string>}[];
};
const bytes = (s: string): Bytes => Uint8Array.from(Buffer.from(s,"hex"));
const join = (...parts: Bytes[]): Bytes => Uint8Array.from(Buffer.concat(parts));
const ascii = (s: string): Bytes => new TextEncoder().encode(s);
function uint(n: bigint, width: number): Bytes {
  if (n < 0n || n >= 1n << BigInt(8*width)) throw new Error("integer overflow");
  const b = new Uint8Array(8); new DataView(b.buffer).setBigUint64(0,n); return b.slice(8-width);
}
function head(major: number, n: bigint): Bytes {
  if (n < 24n) return Uint8Array.of(major*32+Number(n));
  for (const [width,ai] of [[1,24],[2,25],[4,26],[8,27]] as const) if (n < 1n << BigInt(width*8)) return join(Uint8Array.of(major*32+ai),uint(n,width));
  throw new Error("CBOR overflow");
}
function encode(name: string, profile: string, values: Values, unsigned = false): Bytes {
  const def=reg[name]; if (!def) throw new Error("missing schema");
  const fields=Object.entries(def.fields).sort(([a],[b])=>Number(a)-Number(b)).filter(([id])=>!unsigned||Number(id)!==def.mac_field);
  return join(head(5,BigInt(fields.length)),...fields.flatMap(([id,f])=>{
    const v=values[f.name]; let b: Bytes;
    if (f.type.startsWith("uint")) {
      if (typeof v!=="bigint"||v<0n||v>=1n<<BigInt(f.type.slice(4))||(f.const!==undefined&&v!==BigInt(f.const))) throw new Error("integer field");
      b=head(0,v);
    } else if (f.type==="bytes") {
      const length=f.profile_public_key?records.profiles[profile]?.dh_public_bytes:f.length;
      if (!(v instanceof Uint8Array)||v.length!==length) throw new Error("byte field");
      b=join(head(2,BigInt(v.length)),v);
    } else if (f.type==="array") {
      if (!Array.isArray(v)||!f.items) throw new Error("array field");
      b=join(head(4,BigInt(v.length)),...v.map(item=>encode(f.items!.schema_ref,profile,item)));
    } else throw new Error("unsupported field");
    return [head(0,BigInt(id)),b];
  }));
}
function domain(name: string, c: Context, phase=0, role=0, extra: Record<string,Bytes> = {}): Bytes {
  const spec=domains.find(d=>d.name===name); if (!spec) throw new Error("domain");
  const values: Record<string,Bytes>={profile:ascii(c.profile),handshake_hash:bytes(c.handshake_hash_hex),context_digest:bytes(c.context_digest_hex),rekey_id:bytes(c.rekey_id_hex),...extra};
  const ints: Record<string,number>={epoch:c.epoch,next_epoch:c.next_epoch,phase,role};
  return join(bytes(spec.label_bytes),...spec.input_schema.parts.map(p=>{
    if (p.encoding.startsWith("lp-")) {const v=values[p.name];if (!v||(p.length!==undefined&&v.length!==p.length)) throw new Error("domain bytes");return join(uint(BigInt(v.length),4),v);}
    const width=({u8:1,u32:4,u64:8} as Record<string,number>)[p.encoding];
    if (!width||ints[p.name]===undefined) throw new Error("domain integer");return uint(BigInt(ints[p.name]!),width);
  }));
}
function message(c: Context,p: Phase,base: Bytes,values: Values) {
  if (c.next_epoch!==c.epoch+1||c.next_epoch>0xffffffff||base.length!==32) throw new Error("epoch/base");
  const v={...values,phase:BigInt(p.phase),next_epoch:BigInt(c.next_epoch),rekey_id:bytes(c.rekey_id_hex)};
  const role=reg[p.schema]!.sender_role, unsigned=encode(p.schema,c.profile,v,true);
  const info=domain("rekey_confirm_key",c,p.phase,role),key=expand(sha256,base,info,32);
  const macMessage=domain("rekey_confirm_mac",c,p.phase,role,{message:unsigned}),mac=hmac(sha256,key,macMessage);
  return {wire:encode(p.schema,c.profile,{...v,confirmation_mac:mac}),unsigned_hex:unsigned,key_info_hex:info,key_hex:key,mac_message_hex:macMessage,confirmation_mac_hex:mac};
}
function layout(fields: Layout, values: Record<string,bigint>): Bytes {
  return join(...fields.map(f=>{const n=f.const!==undefined?BigInt(f.const):values[f.name],width=({uint8:1,uint16_be:2,uint32_be:4,uint64_be:8} as Record<string,number>)[f.type];if(n===undefined||!width||(f.max!==undefined&&n>BigInt(f.max)))throw new Error("layout");return uint(n,width);}));
}
function recordDomain(name: string,c: Context,integers: Record<string,bigint>,values: Record<string,Bytes>): Bytes {
  const d=domains.find(d=>d.name===name)!;
  return join(bytes(d.label_bytes),...d.input_schema.parts.map(p=>{
    const v={profile:ascii(c.profile),handshake_hash:bytes(c.handshake_hash_hex),...values}[p.name];
    if(p.encoding==="raw")return v!;
    if(p.encoding.startsWith("lp-"))return join(uint(BigInt(v!.length),4),v!);
    return uint(integers[p.name]!,({u8:1,u32:4,u64:8} as Record<string,number>)[p.encoding]!);
  }));
}

it("v4.rekey.ts_reference: independently composes DH, full messages, roots and records",()=>{
  expect(corpus.schema_sha256).toBe(transportV4Registry.schemaSHA256);expect(corpus.rounds).toHaveLength(4);
  const previous=new Map<string,string>(), phases=new Map<string,{c:Context;p:Phase;values:Values}>();
  for(const r of corpus.rounds){
    const c=r.context,spec=records.profiles[c.profile]!,x=spec.dh_algorithm===0;
    const cp=x?x25519.getPublicKey(bytes(r.input.client_private_hex)):p256.getPublicKey(bytes(r.input.client_private_hex),false);
    const sp=x?x25519.getPublicKey(bytes(r.input.server_private_hex)):p256.getPublicKey(bytes(r.input.server_private_hex),false);
    const dh=(key:string,pub:Bytes)=>x?x25519.getSharedSecret(bytes(key),pub):p256.getSharedSecret(bytes(key),pub,false).slice(1,33);
    const shared=dh(r.input.client_private_hex,sp);expect(shared.some(b=>b!==0)).toBe(true);expect(shared).toEqual(dh(r.input.server_private_hex,cp));expect(shared).toEqual(bytes(r.dh_hex));
    if(c.epoch>0)expect(r.old_root_hex).toBe(previous.get(c.profile));
    const secret=expand(sha256,bytes(r.old_root_hex),domain("rekey_secret",c),32);expect(secret).toEqual(bytes(r.secret_hex));
    const barrier=(list:Record<string,string>[]):Values[]=>list.map(item=>Object.fromEntries(Object.entries(item).map(([k,v])=>[k,BigInt(v)])));
    const values:Values[]=[{client_ephemeral:cp,client_barrier:barrier(r.input.client_barrier)},{server_ephemeral:sp,server_barrier:barrier(r.input.server_barrier)},{},{}];
    let init:Bytes=new Uint8Array(),reply:Bytes=new Uint8Array(),root:Bytes=new Uint8Array(),t:Bytes=new Uint8Array();
    for(const [i,p] of r.phases.entries()){
      const v=values[i]!;
      if(i===1){v.init_digest=sha256(domain("rekey_init_digest",c,0,0,{init}));expect(v.init_digest).toEqual(bytes(r.init_digest_hex));}
      if(i===2){t=sha256(domain("rekey_transcript",c,0,0,{init,reply}));expect(t).toEqual(bytes(r.transcript_hex));const prk=hmac(sha256,secret,shared);expect(prk).toEqual(bytes(r.prk_hex));root=expand(sha256,prk,domain("rekey_root",c,0,0,{transcript_digest:t}),32);expect(root).toEqual(bytes(r.new_root_hex));}
      const oldSequence=BigInt(p.role===0?r.input.client_old_sequence:r.input.server_old_sequence);
      if(i>=2){v.transcript_digest=t;v.old_maintenance_next_sequence=oldSequence+1n;}
      const base=i<2?secret:root,m=message(c,p,base,v);expect(base).toEqual(bytes(p.base_hex));expect(m.wire).toEqual(bytes(p.message_hex));
      for(const field of ["unsigned_hex","key_info_hex","key_hex","mac_message_hex","confirmation_mac_hex"] as const)expect(m[field],p.id).toEqual(bytes(p[field]));
      if(i===0)init=m.wire;if(i===1)reply=m.wire;phases.set(p.id,{c,p,values:v});
      const epoch=i<2?c.epoch:c.next_epoch,sequence=i<2?oldSequence:0n,ints={epoch:BigInt(epoch),sequence,direction:BigInt(p.role),sequence_scope:0n};
      expect(p.record.epoch).toBe(epoch);expect(p.record.direction).toBe(reg[p.schema]!.sender_role);expect(p.record.sequence).toBe(String(sequence));
      const header=layout(records.header,ints),nonce=layout(records.nonce,ints),envelope=layout(records.envelope.layout,{payload_length:BigInt(header.length+m.wire.length+spec.tag_bytes),frame_type:BigInt(transportV4Registry.frameTypes.REKEY)});
      const key=expand(sha256,i<2?bytes(r.old_root_hex):root,recordDomain("record_key",c,ints,{}),32),aad=recordDomain("record_aad",c,ints,{envelope_header:envelope,record_header:header});
      const cipher=()=>x?chacha20poly1305(key,nonce,aad):gcm(key,nonce,aad),sealed=cipher().encrypt(m.wire);
      expect(join(envelope,header,sealed)).toEqual(bytes(p.record.wire_hex));expect(cipher().decrypt(sealed)).toEqual(m.wire);
    }
    previous.set(c.profile,r.new_root_hex);
  }
  expect(corpus.negatives.length).toBeGreaterThan(0);
  for(const n of corpus.negatives){
    const source=phases.get(n.source)!;expect(source).toBeDefined();const v={...source.values};
    for(const [k,value] of Object.entries(n.expected))v[k]=k==="old_maintenance_next_sequence"?BigInt(value):bytes(value);
    let matched=false;
    try{matched=Buffer.from(message({...source.c,...n.context_patch},source.p,bytes(n.base_hex),v).wire).equals(Buffer.from(bytes(n.message_hex)));}catch{ /* Invalid expected context has no admissible encoding. */ }
    expect(matched,n.id).toBe(false);
  }
});

File: flowersec-go/internal/protocolv4/rekey_reference_test.go
package protocolv4

// Public fixed transactions only. Exact expected-message matching is not a
// general receiver codec, barrier proof, epoch installation or publisher gate.
import (
	"bytes"
	"crypto/hkdf"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"maps"
	"os"
	"sort"
	"strconv"
	"strings"
	"testing"
)

type rekeyDefinition struct {
	Role   uint64 `json:"sender_role"`
	MAC    uint64 `json:"mac_field"`
	Fields map[string]struct {
		Name, Type    string
		Length        int
		Const         *uint64
		ProfilePublic bool `json:"profile_public_key"`
		Items         struct {
			Schema string `json:"schema_ref"`
		}
	}
}
type rekeyContext struct {
	Profile string `json:"profile"`
	Hash    string `json:"handshake_hash_hex"`
	Digest  string `json:"context_digest_hex"`
	Epoch   uint64 `json:"epoch"`
	Next    uint64 `json:"next_epoch"`
	ID      string `json:"rekey_id_hex"`
}
type rekeyPhase struct {
	ID, Schema  string
	Phase, Role uint64
	Base        string `json:"base_hex"`
	Message     string `json:"message_hex"`
	Unsigned    string `json:"unsigned_hex"`
	KeyInfo     string `json:"key_info_hex"`
	Key         string `json:"key_hex"`
	MACMessage  string `json:"mac_message_hex"`
	MAC         string `json:"confirmation_mac_hex"`
	Expected    map[string]string
	Record      recordVector
}
type rekeyRound struct {
	ID         string
	Context    rekeyContext
	Input      map[string]json.RawMessage
	Phases     []rekeyPhase
	Old        string `json:"old_root_hex"`
	New        string `json:"new_root_hex"`
	Secret     string `json:"secret_hex"`
	DH         string `json:"dh_hex"`
	PRK        string `json:"prk_hex"`
	InitDigest string `json:"init_digest_hex"`
	Transcript string `json:"transcript_hex"`
}

func rekeyEncode(reg map[string]rekeyDefinition, name, profile string, values map[string]any, unsigned bool) ([]byte, error) {
	def, ok := reg[name]
	if !ok {
		return nil, fmt.Errorf("missing map")
	}
	ids := []int{}
	for key := range def.Fields {
		n, err := strconv.Atoi(key)
		if err != nil {
			return nil, err
		}
		if !unsigned || uint64(n) != def.MAC {
			ids = append(ids, n)
		}
	}
	sort.Ints(ids)
	out := readyHead(5, uint64(len(ids)))
	for _, id := range ids {
		f := def.Fields[strconv.Itoa(id)]
		value, ok := values[f.Name]
		if !ok {
			return nil, fmt.Errorf("missing field %s", f.Name)
		}
		out = append(out, readyHead(0, uint64(id))...)
		switch {
		case strings.HasPrefix(f.Type, "uint"):
			n, ok := value.(uint64)
			bits, err := strconv.Atoi(strings.TrimPrefix(f.Type, "uint"))
			if !ok || err != nil || bits < 64 && n >= uint64(1)<<bits || f.Const != nil && n != *f.Const {
				return nil, fmt.Errorf("integer field")
			}
			out = append(out, readyHead(0, n)...)
		case f.Type == "bytes":
			b, ok := value.([]byte)
			length := f.Length
			if f.ProfilePublic {
				_, length, _ = dhReferenceCurve(profile)
			}
			if !ok || len(b) != length {
				return nil, fmt.Errorf("byte field")
			}
			out = append(out, readyHead(2, uint64(len(b)))...)
			out = append(out, b...)
		case f.Type == "array":
			list, ok := value.([]map[string]any)
			if !ok {
				return nil, fmt.Errorf("barrier array")
			}
			out = append(out, readyHead(4, uint64(len(list)))...)
			for _, item := range list {
				b, err := rekeyEncode(reg, f.Items.Schema, profile, item, false)
				if err != nil {
					return nil, err
				}
				out = append(out, b...)
			}
		default:
			return nil, fmt.Errorf("unsupported field")
		}
	}
	return out, nil
}

func rekeyDomain(c rekeyContext, name string, phase, role uint64, extra map[string][]byte) ([]byte, error) {
	b := map[string][]byte{"profile": []byte(c.Profile)}
	for name, s := range map[string]string{"handshake_hash": c.Hash, "context_digest": c.Digest, "rekey_id": c.ID} {
		v, err := hex.DecodeString(s)
		if err != nil {
			return nil, err
		}
		b[name] = v
	}
	maps.Copy(b, extra)
	return recordDomain(name, b, map[string]uint64{"epoch": c.Epoch, "next_epoch": c.Next, "phase": phase, "role": role})
}
func rekeyExpand(key, info []byte) []byte {
	b, err := hkdf.Expand(sha256.New, key, string(info), 32)
	if err != nil {
		panic(err)
	}
	return b
}
func rekeyHMAC(key, info []byte) []byte {
	h := hmac.New(sha256.New, key)
	_, _ = h.Write(info)
	return h.Sum(nil)
}
func rekeyHash(b []byte) []byte { h := sha256.Sum256(b); return h[:] }

func rekeyMessage(reg map[string]rekeyDefinition, c rekeyContext, p rekeyPhase, base []byte, values map[string]any) ([]byte, map[string][]byte, error) {
	if c.Next != c.Epoch+1 || c.Next > uint64(^uint32(0)) || len(base) != 32 {
		return nil, nil, fmt.Errorf("epoch/base")
	}
	def := reg[p.Schema]
	values = maps.Clone(values)
	values["phase"] = p.Phase
	values["next_epoch"] = c.Next
	var err error
	values["rekey_id"], err = hex.DecodeString(c.ID)
	if err != nil {
		return nil, nil, err
	}
	u, err := rekeyEncode(reg, p.Schema, c.Profile, values, true)
	if err != nil {
		return nil, nil, err
	}
	info, err := rekeyDomain(c, "rekey_confirm_key", p.Phase, def.Role, nil)
	if err != nil {
		return nil, nil, err
	}
	key := rekeyExpand(base, info)
	msg, err := rekeyDomain(c, "rekey_confirm_mac", p.Phase, def.Role, map[string][]byte{"message": u})
	if err != nil {
		return nil, nil, err
	}
	mac := rekeyHMAC(key, msg)
	values["confirmation_mac"] = mac
	wire, err := rekeyEncode(reg, p.Schema, c.Profile, values, false)
	return wire, map[string][]byte{"unsigned_hex": u, "key_info_hex": info, "key_hex": key, "mac_message_hex": msg, "confirmation_mac_hex": mac}, err
}

func TestV4RekeySharedCorpus(t *testing.T) {
	var corpus struct {
		Schema    string `json:"schema_sha256"`
		Rounds    []rekeyRound
		Negatives []struct {
			ID, Source string
			Message    string                     `json:"message_hex"`
			Base       string                     `json:"base_hex"`
			Patch      map[string]json.RawMessage `json:"context_patch"`
			Expected   map[string]string
		}
	}
	raw, err := os.ReadFile("../../../testdata/transport_v4/rekey.json")
	if err != nil {
		t.Fatal(err)
	}
	if err = json.Unmarshal(raw, &corpus); err != nil {
		t.Fatal(err)
	}
	if corpus.Schema != SchemaSHA256 || len(corpus.Rounds) != 4 || len(corpus.Negatives) == 0 {
		t.Fatal("rekey corpus binding")
	}
	var reg map[string]rekeyDefinition
	if err = json.Unmarshal([]byte(RekeyRegistryJSON), &reg); err != nil {
		t.Fatal(err)
	}
	var records struct {
		Envelope      struct{ Layout []recordField }
		Header, Nonce []recordField
		Profiles      map[string]struct {
			Algorithm string `json:"record_aead"`
			Tag       uint64 `json:"tag_bytes"`
		}
	}
	if err = json.Unmarshal([]byte(RecordRegistryJSON), &records); err != nil {
		t.Fatal(err)
	}
	type match struct {
		context rekeyContext
		phase   rekeyPhase
		values  map[string]any
	}
	byID := map[string]match{}
	previous := map[string]string{}
	for _, r := range corpus.Rounds {
		c := r.Context
		check := func(actual []byte, want string) {
			t.Helper()
			if hex.EncodeToString(actual) != want {
				t.Fatalf("%s bytes differ", r.ID)
			}
		}
		get := func(name string) string {
			var s string
			if err := json.Unmarshal(r.Input[name], &s); err != nil {
				t.Fatal(err)
			}
			return s
		}
		b := func(s string) []byte { return noiseHex(t, s) }
		domain := func(name string, extra map[string][]byte) []byte {
			v, err := rekeyDomain(c, name, 0, 0, extra)
			if err != nil {
				t.Fatal(err)
			}
			return v
		}
		if c.Epoch > 0 && previous[c.Profile] != r.Old {
			t.Fatal("root continuity")
		}
		cp, err := dhPublicReference(c.Profile, b(get("client_private_hex")))
		if err != nil {
			t.Fatal(err)
		}
		sp, err := dhPublicReference(c.Profile, b(get("server_private_hex")))
		if err != nil {
			t.Fatal(err)
		}
		dh, err := dhReference(c.Profile, b(get("client_private_hex")), sp)
		if err != nil {
			t.Fatal(err)
		}
		check(dh, r.DH)
		other, err := dhReference(c.Profile, b(get("server_private_hex")), cp)
		if err != nil || !bytes.Equal(dh, other) {
			t.Fatal("DH directions", err)
		}
		secret := rekeyExpand(b(r.Old), domain("rekey_secret", nil))
		check(secret, r.Secret)
		barrier := func(side string) []map[string]any {
			var entries []map[string]string
			if err := json.Unmarshal(r.Input[side+"_barrier"], &entries); err != nil {
				t.Fatal(err)
			}
			out := []map[string]any{}
			for _, e := range entries {
				item := map[string]any{}
				for k, v := range e {
					n, err := strconv.ParseUint(v, 10, 64)
					if err != nil {
						t.Fatal(err)
					}
					item[k] = n
				}
				out = append(out, item)
			}
			return out
		}
		values := []map[string]any{{"client_ephemeral": cp, "client_barrier": barrier("client")}, {"server_ephemeral": sp, "server_barrier": barrier("server")}, {}, {}}
		var init, reply, root, transcript []byte
		for i, p := range r.Phases {
			base := secret
			if i == 1 {
				values[i]["init_digest"] = rekeyHash(domain("rekey_init_digest", map[string][]byte{"init": init}))
				check(values[i]["init_digest"].([]byte), r.InitDigest)
			}
			if i == 2 {
				transcript = rekeyHash(domain("rekey_transcript", map[string][]byte{"init": init, "reply": reply}))
				check(transcript, r.Transcript)
				prk := rekeyHMAC(secret, dh)
				check(prk, r.PRK)
				root = rekeyExpand(prk, domain("rekey_root", map[string][]byte{"transcript_digest": transcript}))
				check(root, r.New)
			}
			if i >= 2 {
				base = root
				values[i]["transcript_digest"] = transcript
				side := "client"
				if p.Role == 1 {
					side = "server"
				}
				seq, err := strconv.ParseUint(get(side+"_old_sequence"), 10, 64)
				if err != nil {
					t.Fatal(err)
				}
				values[i]["old_maintenance_next_sequence"] = seq + 1
			}
			wire, material, err := rekeyMessage(reg, c, p, base, values[i])
			if err != nil {
				t.Fatal(err)
			}
			check(wire, p.Message)
			check(base, p.Base)
			for k, want := range map[string]string{"unsigned_hex": p.Unsigned, "key_info_hex": p.KeyInfo, "key_hex": p.Key, "mac_message_hex": p.MACMessage, "confirmation_mac_hex": p.MAC} {
				check(material[k], want)
			}
			if i == 0 {
				init = wire
			}
			if i == 1 {
				reply = wire
			}
			byID[p.ID] = match{c, p, values[i]}
			epoch, recordRoot := c.Epoch, b(r.Old)
			side := "client"
			if p.Role == 1 {
				side = "server"
			}
			sequence, err := strconv.ParseUint(get(side+"_old_sequence"), 10, 64)
			if err != nil {
				t.Fatal(err)
			}
			if i >= 2 {
				epoch, recordRoot, sequence = c.Next, root, 0
			}
			if p.Record.Epoch != epoch || p.Record.Direction != reg[p.Schema].Role || p.Record.Sequence != sequence {
				t.Fatal("record phase")
			}
			ints := map[string]uint64{"epoch": epoch, "direction": p.Role, "sequence_scope": 0, "sequence": sequence}
			header, err := recordLayout(records.Header, ints)
			if err != nil {
				t.Fatal(err)
			}
			nonce, err := recordLayout(records.Nonce, ints)
			if err != nil {
				t.Fatal(err)
			}
			spec := records.Profiles[c.Profile]
			envelope, err := recordLayout(records.Envelope.Layout, map[string]uint64{"payload_length": uint64(len(header)+len(wire)) + spec.Tag, "frame_type": uint64(FrameRekey)})
			if err != nil {
				t.Fatal(err)
			}
			info, err := recordDomain("record_key", map[string][]byte{"profile": []byte(c.Profile), "handshake_hash": b(c.Hash)}, ints)
			if err != nil {
				t.Fatal(err)
			}
			key := rekeyExpand(recordRoot, info)
			aad, err := recordDomain("record_aad", map[string][]byte{"profile": []byte(c.Profile), "envelope_header": envelope, "record_header": header}, ints)
			if err != nil {
				t.Fatal(err)
			}
			aead, err := recordAEAD(spec.Algorithm, key)
			if err != nil {
				t.Fatal(err)
			}
			sealed := aead.Seal(nil, nonce, wire, aad)
			check(bytes.Join([][]byte{envelope, header, sealed}, nil), p.Record.Wire)
			opened, err := aead.Open(nil, nonce, sealed, aad)
			if err != nil || !bytes.Equal(opened, wire) {
				t.Fatal("record authentication", err)
			}
		}
		previous[c.Profile] = r.New
	}
	for _, n := range corpus.Negatives {
		m, ok := byID[n.Source]
		if !ok {
			t.Fatal("negative source")
		}
		raw, err := json.Marshal(m.context)
		if err != nil {
			t.Fatal(err)
		}
		var patched map[string]json.RawMessage
		if err = json.Unmarshal(raw, &patched); err != nil {
			t.Fatal(err)
		}
		maps.Copy(patched, n.Patch)
		raw, err = json.Marshal(patched)
		if err != nil {
			t.Fatal(err)
		}
		var c rekeyContext
		if err = json.Unmarshal(raw, &c); err != nil {
			t.Fatal(err)
		}
		values := maps.Clone(m.values)
		for k, s := range n.Expected {
			if k == "old_maintenance_next_sequence" {
				v, err := strconv.ParseUint(s, 10, 64)
				if err != nil {
					t.Fatal(err)
				}
				values[k] = v
			} else {
				values[k] = noiseHex(t, s)
			}
		}
		wire, _, err := rekeyMessage(reg, c, m.phase, noiseHex(t, n.Base), values)
		if err == nil && bytes.Equal(wire, noiseHex(t, n.Message)) {
			t.Fatal("negative matched", n.ID)
		}
	}
}
